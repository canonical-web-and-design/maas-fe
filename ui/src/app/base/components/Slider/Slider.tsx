import Field from "@canonical/react-components/dist/components/Field";
import React from "react";

const DEFAULT_FILLED_COLOR = "#0066CC";

type Props = {
  disabled?: boolean;
  emptyColor?: string;
  error?: string;
  filledColor?: string;
  help?: string;
  inputDisabled?: boolean;
  label?: string;
  max: number;
  min: number;
  onChange: (e: React.FormEvent<HTMLInputElement>) => void;
  required?: boolean;
  showInput?: boolean;
  step?: number;
  value: number;
};

export const Slider = ({
  disabled = false,
  emptyColor = "transparent",
  error,
  filledColor = DEFAULT_FILLED_COLOR,
  help,
  inputDisabled = false,
  label,
  max,
  min,
  onChange,
  required = false,
  showInput = false,
  step = 1,
  value,
}: Props): JSX.Element => {
  let style = {};
  if (navigator?.userAgent?.includes("AppleWebKit")) {
    // Range inputs on Webkit browsers don't have a built-in "filled" portion,
    // so instead it is handled here as a background.
    const filledPercentage = `${((value - min) / (max - min)) * 100}%`;
    style = {
      background: `linear-gradient(
        to right,
        ${filledColor} 0%,
        ${filledColor} ${filledPercentage},
        ${emptyColor} ${filledPercentage},
        ${emptyColor} 100%
      )`,
    };
  }
  return (
    <Field error={error} help={help} label={label}>
      <div className="p-slider__wrapper">
        <input
          className="p-slider"
          disabled={disabled}
          max={max}
          min={min}
          onChange={onChange}
          required={required}
          step={step}
          style={style}
          type="range"
          value={value}
        />
        {showInput && (
          <input
            className="p-slider__input"
            disabled={disabled || inputDisabled}
            max={max}
            min={min}
            onChange={onChange}
            step={step}
            type="number"
            value={value}
          />
        )}
      </div>
    </Field>
  );
};

export default Slider;
