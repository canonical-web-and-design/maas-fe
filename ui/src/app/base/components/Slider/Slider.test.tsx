import { shallow } from "enzyme";
import React from "react";

import Slider from "./Slider";

describe("Slider", () => {
  it("renders correctly", () => {
    const wrapper = shallow(
      <Slider max={10} min={0} onChange={jest.fn()} showInput value={5} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can render without an additional number input", () => {
    const wrapper = shallow(
      <Slider
        max={10}
        min={0}
        onChange={jest.fn()}
        showInput={false}
        value={5}
      />
    );
    expect(wrapper.find("input[type='number']").exists()).toBe(false);
  });

  it("correctly displays filled portion", () => {
    const wrapper = shallow(
      <Slider max={10} min={2} onChange={jest.fn()} showInput value={4} />
    );
    const backgroundStyle = wrapper.find(".p-slider").props().style
      ?.background as string;
    const trimmed = backgroundStyle?.replace(/\s\s+/g, " ");
    const expectedPercentage = "25%"; // ((4 - 2) / (10 - 2)) * 100 = (2 / 8) * 100 = 25
    expect(trimmed.includes(expectedPercentage)).toBe(true);
  });

  it("can be given a custom filled color", () => {
    const wrapper = shallow(
      <Slider
        filledColor="#abc123"
        max={8}
        min={2}
        onChange={jest.fn()}
        showInput
        value={4}
      />
    );
    const backgroundStyle = wrapper.find(".p-slider").props().style
      ?.background as string;
    const trimmed = backgroundStyle.replace(/\s\s+/g, " ");
    expect(trimmed.includes("#abc123")).toBe(true);
  });
});
