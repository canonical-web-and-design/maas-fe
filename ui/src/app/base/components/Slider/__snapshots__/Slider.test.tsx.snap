// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders correctly 1`] = `
<Field>
  <div
    className="p-slider__wrapper"
  >
    <input
      className="p-slider"
      disabled={false}
      max={10}
      min={0}
      onChange={[MockFunction]}
      required={false}
      step={1}
      style={
        Object {
          "background": "linear-gradient(
        to right,
        #0066CC 0%,
        #0066CC 50%,
        transparent 50%,
        transparent 100%
      )",
        }
      }
      type="range"
      value={5}
    />
    <input
      className="p-slider__input"
      disabled={false}
      max={10}
      min={0}
      onChange={[MockFunction]}
      step={1}
      type="number"
      value={5}
    />
  </div>
</Field>
`;
